### PR TITLE
Awesome DHT PART I

### PR DESCRIPTION
**Pull request by dignifiedquire**
Main work tracker: https://github.com/libp2p/js-libp2p-dht/pull/1

- [ ] Make usage of the DHT efficient (avoid doing multiple query calls for the same file)
  - [ ] add .getMany js-ipfs-bitswap
  - [ ] add .getMany js-ipfs-block-service
  - [ ] use .getMany in unixfs-engine
- [x] interface-ipfs-core
  - [x] implement remaining DHT methods in js-ipfs-api
  - [x] Update tests and spec, ensure that both pass in js-ipfs and js-ipfs-api
  - [ ] Make tests and API better https://github.com/ipfs/interface-ipfs-core/issues/136 (non critical)
- [x] js-libp2p-kad-dht
  - [x] Implement it
  - [x] Integrate in js-libp2p js-libp2p https://github.com/libp2p/js-libp2p/pull/86
  - [x] Integrate in js-libp2p-ipfs-nodejs https://github.com/ipfs/js-libp2p-ipfs-nodejs
  - [ ] ~~Integrate libp2p-ipfs-browser~~ will only add DHT there after Relay is in
  - [ ] Make random walk be a discovery service https://github.com/libp2p/js-libp2p-kad-dht/issues/7
- [ ] Implement .dht in js-ipfs and
  - [x] core api
  - [ ] http-api
  - [ ] cli
- [ ] fix tests for DHT is not available (currently always happen in the browser)
- [ ] interop tests
  - [x] js node in go-ipfs-land (implemented, not green)
  - [ ] go node in js-ipfs-land

**UPDATE:** I (@diasdavid) have been updating and expanding this list as needed

### CODE REVIEW

File: src/core/components/dht.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      self.libp2p.dht.getClosestPeers(peer.id, callback)
+    }),
+    /**
+     * Find peers in the DHT that can provide a specific value, given a key.
+     *
+     * @param {CID} key - They key to find providers for.
+     * @param {function(Error, Array<PeerInfo>)} [callback]
+     * @returns {Promise<PeerInfo>|void}
+     */
+    findprovs: promisify((key, callback) => {
```
Comment (diasdavid) [inline src/core/components/dht.js]: nvm, let's keep using go-ipfs API for now - https://github.com/ipfs/interface-ipfs-core/issues/136

File: src/core/components/dht.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      self.libp2p.dht.findProviders(key, callback)
+    }),
+    /**
+     * Query the dHT for all multiaddresses associated with a `PeerId`.
+     *
+     * @param {PeerId} peer - The id of the peer to search for.
+     * @param {function(Error, Array<Multiaddr>)} [callback]
+     * @returns {Promise<Array<Multiaddr>>|void}
+     */
+    findpeer: promisify((peer, callback) => {
```
Comment (diasdavid) [inline src/core/components/dht.js]: nvm, let's keep using go-ipfs API for now - https://github.com/ipfs/interface-ipfs-core/issues/136

**Comment by dignifiedquire** [comment]
> realised now that I've never submitted the CR :( Don't worry, I am working on it :)

:D

**Comment by diasdavid** [comment]
Ok, here is an update:

- DHT is  integrated with js-ipfs with this PR at the core level
- Good news is that it indeed does DHT things between js-ipfs nodes
- Bad news is that it doesn't interop yet with go-ipfs
- Annoying news is that it breaks silently

**Comment by diasdavid** [comment]
This PR got quite big and caused changed throughout the libp2p interfaces, I was hoping to get this all in one release but it seems that now I'm blocking @pgte and the [IIIF endeavor](https://github.com/ipfs/notes/issues/240) from moving smoothly. 

I'm going to divide this endeavor in Part I and part II, Part I being all the internal pieces that were already developed and integrated and Part II will be the final interop tests and remaining pieces to make the use of DHT faster.

**Pull request merged**